src/PublicAllocator.sol

```diff
@@ -59,36 +65,43 @@ contract PublicAllocator is IPublicAllocatorStaticTyping {
 
     /// PUBLIC ///
 
-    function reallocate(MarketAllocation[] calldata allocations) external payable {
+    function withdrawTo(Withdrawal[] calldata withdrawals, MarketParams calldata depositMarketParams)
+        external
+        payable
+    {
         if (msg.value != fee) revert ErrorsLib.IncorrectFee(msg.value);
 
-        uint256[] memory assets = new uint256[](allocations.length);
-        for (uint256 i = 0; i < allocations.length; ++i) {
-            // Do not compute interest twice for every market
-            MORPHO.accrueInterest(allocations[i].marketParams);
-            assets[i] = MORPHO.expectedSupplyAssets(allocations[i].marketParams, address(VAULT));
-        }
+        MarketAllocation[] memory allocations = new MarketAllocation[](withdrawals.length + 1);
+        allocations[withdrawals.length].marketParams = depositMarketParams;
+        allocations[withdrawals.length].assets = type(uint256).max;
 
-        VAULT.reallocate(allocations);
+        uint128 totalWithdrawn;
 
-        MarketParams memory marketParams;
-        for (uint256 i = 0; i < allocations.length; ++i) {
-            marketParams = allocations[i].marketParams;
-            Id id = marketParams.id();
-            uint256 newAssets = MORPHO.expectedSupplyAssets(marketParams, address(VAULT));
-            if (newAssets >= assets[i]) {
-                if (newAssets > supplyCap[id]) revert ErrorsLib.PublicAllocatorSupplyCapExceeded(id);
-                uint128 inflow = (newAssets - assets[i]).toUint128();
-                flowCap[id].maxIn -= inflow;
-                flowCap[id].maxOut = (flowCap[id].maxOut).saturatingAdd(inflow);
-            } else {
-                uint128 outflow = (assets[i] - newAssets).toUint128();
-                flowCap[id].maxIn = (flowCap[id].maxIn).saturatingAdd(outflow);
-                flowCap[id].maxOut -= outflow;
+        for (uint256 i = 0; i < withdrawals.length; ++i) {
+            allocations[i].marketParams = withdrawals[i].marketParams;
+            Id id = withdrawals[i].marketParams.id();
+            uint256 assets = MORPHO.expectedSupplyAssets(withdrawals[i].marketParams, address(VAULT));
+            uint128 withdrawnAssets = withdrawals[i].amount;
+            // Clamp at 0 if withdrawnAssets is too big
+            if (withdrawnAssets > assets) {
+                withdrawnAssets = assets.toUint128();
             }
+
+            totalWithdrawn += withdrawnAssets;
+            allocations[i].assets = assets - withdrawnAssets;
+            flowCap[id].maxIn = (flowCap[id].maxIn).saturatingAdd(withdrawnAssets);
+            flowCap[id].maxOut -= withdrawnAssets;
+            emit EventsLib.PublicWithdrawal(id, withdrawnAssets);
         }
 
-        emit EventsLib.PublicReallocate(msg.sender, msg.value);
+        VAULT.reallocate(allocations);
+
+        Id depositMarketId = depositMarketParams.id();
+        uint256 depositAssets = MORPHO.expectedSupplyAssets(depositMarketParams, address(VAULT));
```

Comment:
Here we can call supply shares directly and convert to assets, because interest have been accrued during reallocation

Reply:
Right, probably not worth it. Closing this